app/models/user.rb

```diff
@@ -142,6 +142,14 @@ def active?
     is_active
   end
 
+  def available_deputies
+    self
+      .class
+      .joins(:deputy_assignments)
+      .where(deputy_assignments: { primary_user_id: id, is_active: true, deactivated_at: nil })
+      .where.not(deputy_assignments: { confirmed_at: nil })
+  end
```

Comment:
I think this should be memoized, but also I think we should write a more efficient version for use with the masquerading. 

Right now we load out all the available deputies, but we only ever ask "is user X an avaliable deputy of user Y" which requires scanning through the list. We could accomplish the same thing faster with an implementation that just hits the table for the pair that we're interested in:

```ruby
Class User
def active_deputy?(another_user)
  deputy_assignments
    .where(deputy: another_user)
    .active
    .confirmed   # This scope will need to be implemented I think
    .any?
end
```

Reply:
I agree on making it more efficient. I'm not sure about the memoization because this has bitten me before. I'm not always sure when the object gets refreshed and a user could be waiting for a new deputy that's been approved but can be used yet because the data is stale.

Reply:
```
> primary_assignments.where(deputy: other_user).active.confirmed.explain
=> EXPLAIN for: SELECT "deputy_assignments".* FROM "deputy_assignments" WHERE "deputy_assignments"."primary_user_id" = $1 AND "deputy_assignments"."deputy_user_id" = $2 AND "deputy_assignments"."is_active" = $3 AND (CONFIRMED_AT IS NOT NULL) [["primary_user_id", 7], ["deputy_user_id", 2], ["is_active", true]]
                                                          QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------
 Index Scan using index_deputy_assignments_on_unique_users_if_active on deputy_assignments  (cost=0.15..8.17 rows=1 width=65)
   Index Cond: ((primary_user_id = '7'::bigint) AND (deputy_user_id = '2'::bigint))
   Filter: (is_active AND (confirmed_at IS NOT NULL))
(3 rows)
```

versus

```
> self.class.joins(:deputy_assignments).where(deputy_assignments: { primary_user_id: id, is_active: true, deactivated_at: nil }).where.not(deputy_assignments: { confirmed_at: nil }).explain
=> EXPLAIN for: SELECT "users".* FROM "users" INNER JOIN "deputy_assignments" ON "deputy_assignments"."deputy_user_id" = "users"."id" WHERE "deputy_assignments"."primary_user_id" = $1 AND "deputy_assignments"."is_active" = $2 AND "deputy_assignments"."deactivated_at" IS NULL AND "deputy_assignments"."confirmed_at" IS NOT NULL [["primary_user_id", 7], ["is_active", true]]
                                                            QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=0.15..9.18 rows=1 width=1003)
   ->  Seq Scan on users  (cost=0.00..1.00 rows=1 width=1003)
   ->  Index Scan using index_deputy_assignments_on_unique_users_if_active on deputy_assignments  (cost=0.15..8.17 rows=1 width=8)
         Index Cond: ((primary_user_id = '7'::bigint) AND (deputy_user_id = users.id))
         Filter: (is_active AND (deactivated_at IS NULL) AND (confirmed_at IS NOT NULL))
(5 rows)
```